src/librustc_mir/build/mod.rs

```diff
@@ -383,6 +383,29 @@ fn construct_fn<'a, 'gcx, 'tcx, A>(hir: Cx<'a, 'gcx, 'tcx>,
     let source_info = builder.source_info(span);
     let call_site_s = (call_site_scope, source_info);
     unpack!(block = builder.in_scope(call_site_s, LintLevel::Inherited, block, |builder| {
+
+        // According to rustc_trans/callee we mark all foreign instances with nounwind,
+        // so let's make sure we don't unwind.
+        // Therefore generate an extra "Abort" landing pad.
+
+        // FIXME: Figure out why we can't use something like this instead:
+        // tcx.is_foreign_item(tcx.hir.local_def_id(fn_id));
+        // tcx.has_attr(tcx.hir.local_def_id(fn_id), "unwind");
```

Comment:
What happens if you use `has_attr`?

Reply:
AFAICT, `tcx.has_attr(tcx.hir.local_def_id(fn_id), "unwind")` returns `false` also for "__rust_start_panic" and "panicking::rust_begin_panic". 

I don't know why, all these different contexts and ids are a bit bewildering to me. My guess is that I'm trying with the wrong ID or something, but then I don't know what ID would be the right one.

Reply:
Scratch this comment, I must have done something wrong. It does seem to work in the new - just pushed - version.